File: 05_fps_counter/game.c
// This code is based on Gigi's Empty SDL2 window and handling keyboard and 
// mouse events tutorials:
// http://gigi.nullneuron.net/gigilabs/showing-an-empty-window-in-sdl2/
// http://gigi.nullneuron.net/gigilabs/handling-keyboard-and-mouse-events-in-sdl2/

#include <stdbool.h>
#include <string.h>
#include <SDL2/SDL.h>
#include "time.h"

typedef struct {
  float x;
  float y;
  float vx;
  float vy;
  float ax;
  float ay;
  float f;
} Player;

typedef struct {
  double dt;
  Timespec t;
  double acc;
  Timespec fpsUpdate;
} Time;

SDL_Texture * loadTexture(SDL_Renderer * renderer) {
  SDL_Surface * image = SDL_LoadBMP("man.bmp");
  SDL_Texture * texture = SDL_CreateTextureFromSurface(renderer, image);
  SDL_FreeSurface(image);
  return texture;
}

Player initPlayer() {
  Player player = {
    .x = 288,
    .y = 208,
    .vx = 0,
    .vy = 0,
    .ax = 0,
    .ay = 0,
    .f = 1000
  };
  return player;
}

Time initTime() {
  Time time = {
    .dt = 0.01,
    .t = currentTime(),
    .acc = 0,
    .fpsUpdate = currentTime()
  };
  return time;
}

void handleKeyboard(Player * player) {
  const Uint8 * keyboardState = SDL_GetKeyboardState(NULL);
  if (keyboardState[SDL_SCANCODE_A] && keyboardState[SDL_SCANCODE_D]) {
    player->ax = 0;
  } else if (keyboardState[SDL_SCANCODE_A]) {
    player->ax = -player->f;
  } else if (keyboardState[SDL_SCANCODE_D]) {
    player->ax = player->f;
  } else {
    player->ax = 0;
  }
  if (keyboardState[SDL_SCANCODE_W] && keyboardState[SDL_SCANCODE_S]) {
    player->ay = 0;
  } else if (keyboardState[SDL_SCANCODE_W]) {
    player->ay = -player->f;
  } else if (keyboardState[SDL_SCANCODE_S]) {
    player->ay = player->f;
  } else {
    player->ay = 0;
  }
}

void handleEvents(bool * quit) {
  SDL_Event event;
  SDL_PollEvent(&event);
  switch (event.type)
  {
    case SDL_QUIT:
      *quit = true;
      break;
  }
}

int cleanUpAndExit(SDL_Texture * texture, SDL_Renderer * renderer, SDL_Window * window) {
  SDL_DestroyTexture(texture);
  SDL_DestroyRenderer(renderer);
  SDL_DestroyWindow(window); 
  SDL_Quit();
  return 0;
}

void integrate(Player * player, double dt) {
  player->vx += player->ax * dt;
  player->x += player->vx * dt;

  player->vy += player->ay * dt;
  player->y += player->vy * dt;
}

void render(Player * player, SDL_Texture * texture, SDL_Renderer * renderer) {
  SDL_Rect dstrect = { player->x, player->y, 64, 64 };
  SDL_RenderClear(renderer);
  SDL_RenderCopy(renderer, texture, NULL, &dstrect);
  SDL_RenderPresent(renderer);
}

void fps(Time * time, SDL_Window * window) {
  static int frames = 0;
  double timeSinceLastUpdate = diffTime(currentTime(), time->fpsUpdate);
  if ( timeSinceLastUpdate > 1 ) {
    char caption[80];
    sprintf(caption, "FPS: %d", frames);
    SDL_SetWindowTitle(window, caption);
    time->fpsUpdate = currentTime();
    frames = 0;
  }
  ++frames;
}

int main(int argc, char ** argv)
{
  SDL_Init(SDL_INIT_VIDEO);

  bool quit = false;
  SDL_Window * window = SDL_CreateWindow(
      "Game",
      SDL_WINDOWPOS_UNDEFINED,
      SDL_WINDOWPOS_UNDEFINED,
      640,
      480,
      0);
  SDL_Renderer * renderer = SDL_CreateRenderer(window, -1, 0);
  SDL_Texture * texture = loadTexture(renderer);

  Player curPlayer = initPlayer();
  Player prevPlayer;

  Time time = initTime();

  SDL_SetRenderDrawColor(renderer, 255, 255, 255, 255);

  while (!quit)
  {
    Timespec newTime = currentTime();
    double frameTime = diffTime(newTime, time.t);

    if ( frameTime > 0.25 ) {
      frameTime = 0.25;
    }
    time.t = newTime;

    time.acc += frameTime;

    while ( time.acc >= time.dt )
    {
      prevPlayer = curPlayer;
      integrate(&curPlayer, time.dt);
      time.t = timespecAdd(secondsToTimespec(time.dt), time.t);
      time.acc -= time.dt;
    }

    const double alpha = time.acc / time.dt;

    Player interpolatedPlayer = {
      .x = curPlayer.x * alpha + prevPlayer.x * ( 1.0 - alpha),
      .y = curPlayer.y * alpha + prevPlayer.y * ( 1.0 - alpha),
    };

    render(&interpolatedPlayer, texture, renderer);
    handleEvents(&quit);
    handleKeyboard(&curPlayer);
    fps(&time, window);
  }
  return cleanUpAndExit(texture, renderer, window);
}

File: 05_fps_counter/game.fs
decimal
require sdl2.fs
require time.fs

: +f! { addr -- }
  addr f@
  f+
  addr f!
;

struct
  float% field pos-x
  float% field pos-y
  float% field ax
  float% field ay
  float% field vx
  float% field vy
  float% field f
end-struct player%

player% %allot constant player
player% %allot constant interpolated-player
player% %allot value previous-player

: init-player ( -- )
  288e player pos-x f!
  208e player pos-y f!
  0e player ax f!
  0e player ay f!
  0e player vx f!
  0e player vy f!
  5000e player f f!
;

0 value window
0 value renderer
0 value texture

sdl-event% %allot constant event

false value quit-flag

: init-sdl ( -- )
  SDL_INIT_VIDEO sdl-init 0< if
    drop
    cr ." Video Initialization failed: "
    sdl-geterror type cr
    bye
  then
;

: create-window ( -- window )
  s" game" SDL_WINDOWPOS_UNDEFINED SDL_WINDOWPOS_UNDEFINED 640 480 0 sdl-create-window
  dup 0= if                   \ error out if not successful
    drop
    cr ." Could not create window: "
    sdl-geterror type cr
    sdl-quit
    bye
  then
;

: create-renderer ( -- renderer )
  window -1 0 sdl-create-renderer
  dup 0= if                   \ error out if not successful
    drop
    cr ." Could not create renderer: "
    sdl-geterror type cr
    window sdl-destroy-window
    sdl-quit
    bye
  then
;


: load-texture ( -- texture )
  renderer
  s" man.bmp" sdl-load-bmp 
  { surface -- }
  surface sdl-create-texture-from-surface
  surface sdl-free-surface
  dup 0= if                   \ error out if not successful
    drop
    cr ." Could not create texture: "
    sdl-geterror type cr
    renderer sdl-destroy-renderer
    window sdl-destroy-window
    sdl-quit
    bye
  then
;

: pending-events? ( -- flag )
  event sdl-poll-event
;

: should-quit? ( -- flag )
  quit-flag 0<>
;

: key-pressed? ( keycode keyboard-state -- flag )
  + c@ 0<>
;

: 2-keys-pressed? { keycode1 keycode2 keyboard-state -- flag }
  keycode2 keyboard-state key-pressed?
  keycode1 keyboard-state key-pressed?
  and
;

: handle-x-movement ( keyboard-state -- )
  SDL_SCANCODE_A SDL_SCANCODE_D 
  { keyboard-state a d -- }
  a d keyboard-state 2-keys-pressed?
  if
    0e player ax f!
  else
    a keyboard-state key-pressed?
    if
      player f f@ fnegate player ax f!
    else
      d keyboard-state key-pressed?
      if 
        player f f@ player ax f!
      else
        0e player ax f!
      endif
    endif
  endif
;

: handle-y-movement ( keyboard-state -- )
  dup SDL_SCANCODE_W SDL_SCANCODE_S rot 2-keys-pressed?
  if
    0e player ay f!
    drop
  else
    dup SDL_SCANCODE_W key-pressed?
    if
      player f f@ fnegate player ay f!
      drop
    else
      SDL_SCANCODE_S key-pressed?
      if 
        player f f@ player ay f!
      else
        0e player ay f!
      endif
    endif
  endif
;

: handle-keyboard ( -- )
  0 sdl-get-keyboard-state
  dup handle-x-movement
  handle-y-movement
;

: handle-events ( -- )
  begin
    pending-events?
  while
    event sdl-event-type ul@
    case
      SDL_QUIT of     \ window close box clicked, or ALT-F4 pressed
        true to quit-flag
      endof
    endcase
  repeat
  handle-keyboard
;

sdl-rect% %allot constant 'player-rect
: player-rect ( -- 'player-rect ) 
  interpolated-player pos-x f@ fround f>s 'player-rect x !
  interpolated-player pos-y f@ fround f>s 'player-rect y !
  64 'player-rect w !
  64 'player-rect h !
  'player-rect
;

: render-player ( -- ) 
  renderer
  texture
  0
  player-rect
  sdl-render-copy drop
;

: render ( -- ) 
  renderer sdl-render-clear drop
  render-player 
  renderer sdl-render-present
;

: set-background-color ( -- ) 
  renderer 255 255 255 255 sdl-set-render-draw-color drop
;

: integrate ( -- )
  begin
    acc@ dt@ f>=
  while
    player to previous-player

    player ax f@ dt@ f* player vx +f!
    player vx f@ dt@ f* player pos-x +f!

    player ay f@ dt@ f* player vy +f!
    player vy f@ dt@ f* player pos-y +f!

    t@ dt@ f+ t!
    acc@ dt@ f- acc!
  repeat
;

: graceful-quit ( -- )
  texture sdl-destroy-texture
  renderer sdl-destroy-renderer
  window sdl-destroy-window
  sdl-quit
;

: interpolate-player-pos-x { F: alpha -- }
  1e alpha f- previous-player pos-x f@ f*
  alpha player pos-x f@ f*
  f+
;

: interpolate-player-pos-y { F: alpha -- }
  1e alpha f- previous-player pos-y f@ f*
  alpha player pos-y f@ f*
  f+
;

: interpolate-player
  acc@ dt@ f/ ( -- alpha )
  fdup ( -- alpha alpha )
  interpolate-player-pos-x interpolated-player pos-x f!
  interpolate-player-pos-y interpolated-player pos-y f!
;

2variable fps-update
utime fps-update 2!
0 value frames
: fps
  utime fps-update 2@ d-
  microseconds-to-seconds
  1e f>= if
    window
    s" FPS: " frames s>d <# #s #> s+
    sdl-set-window-title
    utime fps-update 2!
    0 to frames
  endif
  1 frames + to frames
;

: main-loop
  begin
    calculate-frame-time
    integrate
    interpolate-player
    render
    fps
    handle-events
    should-quit?
  until
;

init-sdl 
create-window to window
create-renderer to renderer
load-texture to texture
set-background-color
init-player
init-time
main-loop
graceful-quit
bye
